subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/type/DefaultArtifactTypeRegistry.java

```diff
@@ -54,12 +54,16 @@ public ArtifactTypeContainer create() {
 
     @Override
     public ImmutableAttributes mapAttributesFor(File file) {
-        String extension = Files.getFileExtension(file.getName());
         ImmutableAttributes attributes = ImmutableAttributes.EMPTY;
-        if (artifactTypeDefinitions != null) {
-            attributes = applyForExtension(attributes, extension);
+        if (file.isDirectory()) {
+            attributes = attributesFactory.of(ARTIFACT_FORMAT, ArtifactTypeDefinition.DIRECTORY);
+        } else {
+            String extension = Files.getFileExtension(file.getName());
+            if (artifactTypeDefinitions != null) {
+                attributes = applyForExtension(attributes, extension);
+            }
+            attributes = attributesFactory.concat(attributesFactory.of(ARTIFACT_FORMAT, extension), attributes);
         }
-        attributes = attributesFactory.concat(attributesFactory.of(ARTIFACT_FORMAT, extension), attributes);
```

Comment:
Can you see if we can add some coverage to DefaultArtifactTypeRegistryTest for this?